cli/vscodeipc/vscodeipc.go

```diff
@@ -0,0 +1,268 @@
+				RegionName: fmt.Sprintf("Unnamed %d", regionID),
+			}
+		}
+		// Convert the microseconds to milliseconds.
+		derpLatency[region.RegionName] = latency * 1000
+	}
+
+	totalRx := uint64(0)
+	totalTx := uint64(0)
+	for _, stat := range api.agentConn.ExtractTrafficStats() {
+		totalRx += stat.RxBytes
+		totalTx += stat.TxBytes
+	}
+	dur := time.Since(api.lastNetwork)
+
+	uploadSecs := float64(totalTx) / dur.Seconds()
+	downloadSecs := float64(totalRx) / dur.Seconds()
+
+	api.lastNetwork = time.Now()
+
+	httpapi.Write(r.Context(), w, http.StatusOK, NetworkResponse{
+		P2P:              p2p,
+		Latency:          float64(latency.Microseconds()) / 1000,
+		PreferredDERP:    derpMap.Regions[node.PreferredDERP].RegionName,
+		DERPLatency:      derpLatency,
+		UploadBytesSec:   int64(uploadSecs),
+		DownloadBytesSec: int64(downloadSecs),
+	})
+}
+
+type ExecuteRequest struct {
+	Command string `json:"command"`
+}
+
+type ExecuteResponse struct {
+	Data     string `json:"data"`
+	ExitCode *int   `json:"exit_code"`
+}
+
+// execute runs the command provided, streams the output back, and returns the exit code.
+func (api *api) execute(w http.ResponseWriter, r *http.Request) {
+	var req ExecuteRequest
+	if !httpapi.Read(r.Context(), w, r, &req) {
+		return
+	}
+	api.sshClientOnce.Do(func() {
+		// The SSH client is lazily created because it's not needed for
+		// all requests. It's only needed for the execute endpoint.
+		api.sshClient, api.sshClientErr = api.agentConn.SSHClient(context.Background())
+	})
+	if api.sshClientErr != nil {
+		httpapi.Write(r.Context(), w, http.StatusInternalServerError, codersdk.Response{
+			Message: "Failed to create SSH client.",
+			Detail:  api.sshClientErr.Error(),
+		})
+		return
+	}
+	session, err := api.sshClient.NewSession()
+	if err != nil {
+		httpapi.Write(r.Context(), w, http.StatusInternalServerError, codersdk.Response{
+			Message: "Failed to create SSH session.",
+			Detail:  err.Error(),
+		})
+		return
+	}
+	defer session.Close()
+	f, ok := w.(http.Flusher)
+	if !ok {
+		panic("http.ResponseWriter is not http.Flusher")
+	}
+
+	execWriter := &execWriter{w, f}
+	session.Stdout = execWriter
+	session.Stderr = execWriter
+	session.Stdin = strings.NewReader(req.Command + "\n")
+	err = session.Start("sh")
+	if err != nil {
+		httpapi.Write(r.Context(), w, http.StatusInternalServerError, codersdk.Response{
+			Message: "Failed to start SSH session.",
+			Detail:  err.Error(),
+		})
+		return
+	}
+	err = session.Wait()
+
+	writeExit := func(exitCode int) {
+		data, _ := json.Marshal(&ExecuteResponse{
+			ExitCode: &exitCode,
+		})
+		_, _ = w.Write(data)
+		f.Flush()
+	}
+
+	if err != nil {
+		var exitError *ssh.ExitError
+		if errors.As(err, &exitError) {
+			writeExit(exitError.ExitStatus())
+			return
+		}
+	}
+	writeExit(0)
+}
+
+type execWriter struct {
+	w http.ResponseWriter
+	f http.Flusher
+}
+
+func (e *execWriter) Write(data []byte) (int, error) {
+	js, err := json.Marshal(&ExecuteResponse{
```

Comment:
nit: is JSON the only option here or maybe we can use a format with a smaller overhead?

Reply:
true

Reply:
I thought about making it binary (just prefixing with a 0 if data and 1 if exit code), but figured it ultimately might not be worth it.

There is very little data that flows through here (in the hundreds of bytes in sum, at most 1000), so the overhead should be pretty minimal as-is. If I'm overlooking what you were proposing, just let me know!